### ISSUE TITLE
Port for CE3

### ISSUE THREAD

**Issue by joroKr21** [zio/interop-cats#260]
Cats-effect 3 final will be out soon including a major redesign of the type classes.
I am working on this, but we will probably need another branch to be able to also maintain the CE2 interop.

**Comment by Fristi** [comment]
Was having the same question, wouldn't it be an idea to have multiple modules? One for CE2 and one for CE3 ?

**Comment by joroKr21** [comment]
Multiple modules or multiple branches - it's definitely easier to update the existing code than start a new module.

**Comment by Fristi** [comment]
That's true, yet I guess there will be some time that some codebases which use this CE2 module in production need time to move over to newer versions of http4s, doobie and fs2 where in time CE3 will be introduced.

**Comment by joroKr21** [comment]
Yes, this can be achieved in two ways - separate branches or separate modules in master.

**Comment by KadekM** [comment]
With cats ecosystem moving forward this would be very nice to have :-)

**Comment by adamgfraser** [comment]
@KadekM #263 by @joroKr21 addresses this.

@joroKr21 How is that going? Do you need any help to get it over the finish line?

**Comment by joroKr21** [comment]
Any help is appreciated - there are some tests failing and others are commented out.
We need to find the causes and fix them.

**Comment by joroKr21** [comment]
Also we need to somehow forward the `Clock` in `ZEnv` to `Ticker` - idk how to do that

**Comment by adamgfraser** [comment]
@joroKr21 You can do that like this:

```scala
def environment(implicit ticker: Ticker): ZEnv = {
  val testClock = new Clock.Service {
    def currentTime(unit: TimeUnit): UIO[Long] =
      ZIO.effectTotal(unit.convert(Duration.fromScala(ticker.ctx.now())))
    val currentDateTime: IO[DateTimeException, OffsetDateTime] =
      currentTime(TimeUnit.MILLISECONDS).map { millis =>
        OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneId.of("UTC"))
      }
    val nanoTime: UIO[Long] =
      ZIO.effectTotal(ticker.ctx.now().toNanos)
    def sleep(duration: Duration): UIO[Unit] =
      duration.asScala match {
        case finiteDuration: FiniteDuration =>
          ZIO.effectAsyncInterrupt { cb =>
            val canceler = ticker.ctx.schedule(finiteDuration, () => cb(UIO.unit))
            Left(UIO.effectTotal(canceler()))
          }
        case infinite: Infinite =>
          ZIO.dieMessage(s"Unexpected infinite duration $infinite passed to Ticker")
      }
  }
  ZEnv.Services.live ++ Has(testClock)
}
```

**Comment by joroKr21** [comment]
@adamgfraser that doesn't work for Scala.js
I get `Referring to non-existent method java.util.concurrent.TimeUnit$$anon$5.convert(java.time.Duration)long`

<details>

```
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$6.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$2.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$4.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$1.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$7.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$5.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit$$anon$3.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
[error] Referring to non-existent method java.util.concurrent.TimeUnit.convert(java.time.Duration)long
[error]   called from private zio.interop.catzSpecBase$$anon$1.$anonfun$currentTime$1(java.util.concurrent.TimeUnit)long
[error]   called from zio.interop.catzSpecBase$$anon$1.currentTime(java.util.concurrent.TimeUnit)zio.ZIO
[error]   called from constructor zio.interop.catzSpecBase$$anon$1.<init>(zio.interop.catzSpecBase,cats.effect.testkit.TestInstances$Ticker)void
[error]   called from zio.interop.catzSpecBase.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from zio.interop.catzSpec.environment(cats.effect.testkit.TestInstances$Ticker)zio.Has
[error]   called from private zio.interop.catzSpecBase.$anonfun$arbitraryEnvironment$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecBase.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryEnvironment(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpecZIOBase.$anonfun$arbitraryClockAndBlocking$1(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Gen
[error]   called from zio.interop.catzSpecZIOBase.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from zio.interop.catzSpec.arbitraryClockAndBlocking(cats.effect.testkit.TestInstances$Ticker)org.scalacheck.Arbitrary
[error]   called from private zio.interop.catzSpec.$anonfun$new$1(cats.effect.testkit.TestInstances$Ticker)org.typelevel.discipline.Laws$RuleSet
[error]   called from constructor zio.interop.catzSpec.<init>()void
[error]   called from static constructor zio.interop.catzSpec.<stinit>()void
[error]   called from core module analyzer
[error] involving instantiated classes:
[error]   zio.interop.catzSpecBase$$anon$1
[error]   zio.interop.catzSpec
```

</details>

**Comment by adamgfraser** [comment]
@joroKr21 You need a `java.time` dependency.

**Comment by joroKr21** [comment]
We have `scala-java-time` as a dependency or did you mean something else? 🤔

**Comment by adamgfraser** [comment]
Hmmm, isn't that the same method that is used [here](https://github.com/zio/zio/blob/e0a2da7b0f026dee7f9812da91dc5cfb1456e112/core/shared/src/main/scala/zio/clock/package.scala#L60)?

**Comment by joroKr21** [comment]
> Hmmm, isn't that the same method that is used here?

Looks like it's an overload. But I have bad news, this doesn't fix any of the test failures 😞 
I think we have a lot of debugging ahead of us.

**Comment by adamgfraser** [comment]
Hmmm, some of these failures do not make sense to me. Here is the simplest one which is failing but seems to pass when I test it directly.

```scala
import zio._
import zio.test._
import zio.test.Assertion._

object InteropSpec extends DefaultRunnableSpec {

  // (F.async[A](k => F.delay(k(Left(e))) >> fu.as(None)) <* F.unit) <-> (F.uncancelable(_ => fu) >> F.raiseError(e))

  val fu = ZIO.unit
  val e  = new Exception("fail")

  def async[A](k: (Either[Throwable, A] => Unit) => Task[Option[Task[Unit]]]): Task[A] =
    ZIO.effectAsyncM(register => k(register.compose(ZIO.fromEither(_))))

  def spec: ZSpec[Environment, Failure] =
    suite("InteropSpec")(
      testM("async.async left is uncancelable sequenced raiseError") {
        val left  = async[Int](k => ZIO.effect(k(Left(e))) *> ZIO.unit.as(None)) <* ZIO.unit
        val right = ZIO.uninterruptibleMask(_ => fu) *> ZIO.fail(e)
        for {
          left  <- left.run
          right <- right.run
        } yield assert(left)(equalTo(right))
      } @@ TestAspect.nonFlaky
    )
}
```

**Comment by joroKr21** [comment]
Maybe there is something else that is wrong in the testing infrastructure - it's the most finicky part so I wouldn't be surprised

**Comment by joroKr21** [comment]
I suspect `toEffect` and in general the fact that we're going from `ZIO` to `IO` for equality - maybe you know a better way.

**Comment by adamgfraser** [comment]
Why don't we just unsafely run them and compare their `Exit` values?

**Comment by joroKr21** [comment]
> Why don't we just unsafely run them and compare their Exit values?

Can't do that on JavaScript right? I think trying to run then just throws an exception.

**Comment by adamgfraser** [comment]
https://github.com/typelevel/cats-effect/blob/71eaad83cea85f7f3751638193380b0cca9b3a49/testkit/shared/src/main/scala/cats/effect/testkit/TestInstances.scala#L165

**Comment by joroKr21** [comment]
I tried to emulate it in dfd488b and now even more tests are failing

**Comment by adamgfraser** [comment]
I think there is something strange going on with the testing infrastructure. For example this works fine but the seemingly identical test fails.

```scala
import cats.effect.testkit._
import zio._
import zio.internal.{Executor, Platform}

import scala.concurrent.duration._

object InteropSpec extends scala.App with TestInstances {

  // (F.async[A](k => F.delay(k(Left(e))) >> fu.as(None)) <* F.unit) <-> (F.uncancelable(_ => fu) >> F.raiseError(e))

  val fu = ZIO.unit
  val e  = new Exception("fail")

  def async[A](k: (Either[Throwable, A] => Unit) => Task[Option[Task[Unit]]]): Task[A] =
    ZIO.effectAsyncM(register => k(register.compose(ZIO.fromEither(_))))

  val left  = async[Int](k => ZIO.effect(k(Left(e))) *> ZIO.unit.as(None)) <* ZIO.unit
  val right = ZIO.uninterruptibleMask(_ => fu) *> ZIO.fail(e)

  implicit def runtime(implicit ticker: Ticker): Runtime[Any] = Runtime(
    (),
    Platform
      .fromExecutor(Executor.fromExecutionContext(1024)(ticker.ctx))
      .withReportFailure(_ => ())
  )

  def unsafeRun[A](zio: UIO[A])(implicit ticker: Ticker): Exit[Nothing, Option[A]] =
    try {

      var exit: Exit[Nothing, Option[A]] = null

      runtime(ticker).unsafeRunAsync {
        zio
          .flatMap(ZIO.succeed(_))
          .on(ticker.ctx)
      } {
        case Exit.Success(a)     => exit = Exit.Success(Some(a))
        case Exit.Failure(cause) => exit = Exit.halt(cause)
      }

      ticker.ctx.tickAll(1.seconds)

      exit
    } catch {
      case t: Throwable =>
        t.printStackTrace()
        throw t
    }
  implicit val ticker = Ticker(TestContext())

  val a = unsafeRun(left.either)
  val b = unsafeRun(right.either)

  println(a)
  println(b)
}
```

**Comment by joroKr21** [comment]
Fixed in #263
